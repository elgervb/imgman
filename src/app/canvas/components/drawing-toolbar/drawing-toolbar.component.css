
evb-formcontrol label {
  width: 100%;
}
