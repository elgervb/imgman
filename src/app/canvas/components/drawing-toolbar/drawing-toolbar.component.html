<form [formGroup]="form">

  <evb-formcontrol>
    <label for="enabled">enable</label>
    <input type="checkbox" id="enabled" formControlName="enabled" />
  </evb-formcontrol>

  <ng-container *ngIf="form.get('enabled').value">

    <hr>

    <evb-formcontrol>
      <input type="radio" id="brushtype-pen" formControlName="brushType" [value]="brushType.pen" checked/>
      <label for="brushtype-pen">pen</label>
    </evb-formcontrol>
    <evb-formcontrol>
      <input type="radio" id="brushtype-marker" formControlName="brushType" [value]="brushType.marker"/>
      <label for="brushtype-marker">marker</label>
    </evb-formcontrol>
    <evb-formcontrol>
      <input type="radio" id="brushtype-circular" formControlName="brushType" [value]="brushType.circular"/>
      <label for="brushtype-circular">circular</label>
    </evb-formcontrol>
    <evb-formcontrol>
      <input type="radio" id="brushtype-spray" formControlName="brushType" [value]="brushType.spray"/>
      <label for="brushtype-spray">spray</label>
    </evb-formcontrol>
    <evb-formcontrol>
      <input type="radio" id="brushtype-multipleLines" formControlName="brushType" [value]="brushType.multipleLines"/>
      <label for="brushtype-multipleLines">multipleLines</label>
    </evb-formcontrol>

    <hr>

      <!-- <button (click)="draw.clear()">clear</button> -->
      <evb-formcontrol>
        <label for="opacity">opacity</label>
        <input type="range" id="opacity" min="0.0" max="1.0" step="0.1" value="1.0" placeholder="Opacity" formControlName="opacity" />
      </evb-formcontrol>

      <evb-formcontrol>
        <label for="line-width">line width</label>
        <input type="range" id="line-width" min="1" max="50" step="1" value="4" placeholder="Line width" formControlName="lineWidth"/>
      </evb-formcontrol>

  </ng-container>

</form>
