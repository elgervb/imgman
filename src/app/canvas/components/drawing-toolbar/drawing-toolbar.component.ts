import { Component, EventEmitter, OnDestroy, OnInit, Output } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { BrushType } from 'src/app/imgman/drawing/models';

const MIN_OPACITY = 0.1;

@Component({
  selector: 'app-drawing-toolbar',
  templateUrl: './drawing-toolbar.component.html',
  styleUrls: ['./drawing-toolbar.component.css']
})
export class DrawingToolbarComponent implements OnInit, OnDestroy {

  @Output() readonly changed = new EventEmitter<DrawingSettings>();

  form: FormGroup;
  brushType = BrushType;

  private readonly destroyed$ = new Subject<void>();

  constructor(private formBuilder: FormBuilder) { }

  ngOnInit() {
    this.form = this.formBuilder.group({
      brushType: [BrushType.marker, Validators.required],
      enabled: [false, Validators.required],
      opacity: [1.0, [Validators.required, Validators.min(MIN_OPACITY), Validators.max(1)]],
      lineWidth: [4, [Validators.required, Validators.min(1), Validators.max(100)]]
    });

    this.form.valueChanges
      .pipe(
        takeUntil(this.destroyed$)
      ).subscribe((value: DrawingSettings) => this.changed.emit(value));
  }

  ngOnDestroy() {
    this.destroyed$.next();
    this.destroyed$.complete();
  }
}

export interface DrawingSettings {
  enabled: boolean;
  brushType: BrushType;
  opacity: number;
  lineWidth: number;
}
