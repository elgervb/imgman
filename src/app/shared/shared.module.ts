import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { ToolbarComponent } from './components/toolbar/toolbar.component';
import { EvbValueAccessorDirective } from './directives/evbcontrolvalueaccessor';
import { IsDarkPipe } from './pipes/is-dark.pipe';
import { RgbToStringPipe } from './pipes/rgb-to-string.pipe';
import { SanitizePipe } from './pipes/sanitize.pipe';

@NgModule({
  declarations: [
    EvbValueAccessorDirective,
    SanitizePipe,
    ToolbarComponent,
    IsDarkPipe,
    RgbToStringPipe,
  ],
  imports: [
    CommonModule
  ],
  exports: [
    EvbValueAccessorDirective,
    IsDarkPipe,
    RgbToStringPipe,
    SanitizePipe,
    ToolbarComponent
  ]
})
export class SharedModule { }
