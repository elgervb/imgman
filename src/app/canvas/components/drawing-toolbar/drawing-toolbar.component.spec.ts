import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { SharedModule } from 'src/app/shared/shared.module';

import { DrawingToolbarComponent } from './drawing-toolbar.component';

describe('DrawingToolbarComponent', () => {
  let component: DrawingToolbarComponent;
  let fixture: ComponentFixture<DrawingToolbarComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [
        DrawingToolbarComponent
      ],
      imports: [
        ReactiveFormsModule,
        SharedModule
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(DrawingToolbarComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();

    expect(fixture).toMatchSnapshot();
  });
});
