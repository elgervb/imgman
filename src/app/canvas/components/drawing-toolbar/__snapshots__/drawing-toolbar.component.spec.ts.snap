// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawingToolbarComponent should create 1`] = `
<app-drawing-toolbar
  brushType={[Function Object]}
  changed={[Function EventEmitter]}
  destroyed$={[Function Subject]}
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <evb-formcontrol
      class=""
    >
      <label>
        Enable drawing
      </label>
      <evb-toggle
        class="ng-untouched ng-pristine ng-valid"
        evbcontrol=""
        formcontrolname="enabled"
        ng-reflect-name="enabled"
      />
    </evb-formcontrol>
    
  </form>
</app-drawing-toolbar>
`;
